File: src/domain/dictionary/poe2Bases.ts
import type { BaseTypeMapping } from './types';

const STATIC_BASE_MAPPINGS: BaseTypeMapping[] = [
  // Waystones (尋路石)
  { zh: '尋路石', en: 'Waystone' },
  { zh: '銘刻尋路石', en: 'Inscribed Waystone' },
  { zh: '塔樓尋路石', en: 'Tower Waystone' },

  // Uncut Gems (未切割寶石)
  { zh: '未切割技能寶石', en: 'Uncut Skill Gem' },
  { zh: '未切割精魂寶石', en: 'Uncut Spirit Gem' },
  { zh: '未切割輔助寶石', en: 'Uncut Support Gem' },

  // Runes (符文)
  { zh: '符文', en: 'Rune' },
  { zh: '太陽符文', en: 'Sun Rune' },
  { zh: '鐵符文', en: 'Iron Rune' },
  { zh: '心靈符文', en: 'Mind Rune' },
  { zh: '石之符文', en: 'Stone Rune' },
  { zh: '風暴符文', en: 'Storm Rune' },
  { zh: '冰川符文', en: 'Glacial Rune' },
  { zh: '沙漠符文', en: 'Desert Rune' },
  { zh: '劇毒符文', en: 'Poison Rune' },

  // Currencies & Gold
  { zh: '金幣', en: 'Gold' },
  { zh: '高階重鑄石', en: 'Greater Orb of Scouring' },
  { zh: '次級重鑄石', en: 'Lesser Orb of Scouring' },
  { zh: '高階點金石', en: 'Greater Orb of Alchemy' },
  { zh: '次級點金石', en: 'Lesser Orb of Alchemy' },
  { zh: '重挫石', en: 'Orb of Annulment' },
  { zh: '神聖石', en: 'Divine Orb' },
  { zh: '崇高石', en: 'Exalted Orb' },
  { zh: '混沌石', en: 'Chaos Orb' },
  { zh: '瓦爾寶珠', en: 'Vaal Orb' },
  { zh: '機會石', en: 'Orb of Chance' },
];

export function getPoe2CanonicalBaseMappings(): BaseTypeMapping[] {
  const list = [...STATIC_BASE_MAPPINGS];

  for (let tier = 1; tier <= 16; tier++) {
    list.push({
      zh: `尋路石 (階級 ${tier})`,
      en: `Waystone (Tier ${tier})`,
    });
  }

  for (let tier = 1; tier <= 20; tier++) {
    list.push({
      zh: `未切割寶石 (階級 ${tier})`,
      en: `Uncut Gem (Tier ${tier})`,
    });
  }

  return list;
}

export function getPoe2ItemMap(): Map<string, string> {
  const map = new Map<string, string>();

  for (const { zh, en } of getPoe2CanonicalBaseMappings()) {
    map.set(zh, en);
    map.set(en, en);
  }

  // Shorthand aliases for Waystones and Uncut Gems
  for (let tier = 1; tier <= 16; tier++) {
    const en = `Waystone (Tier ${tier})`;
    map.set(`尋路石 T${tier}`, en);
    map.set(`尋路石 (T${tier})`, en);
  }

  for (let tier = 1; tier <= 20; tier++) {
    const en = `Uncut Gem (Tier ${tier})`;
    map.set(`未切割寶石 T${tier}`, en);
  }

  return map;
}

File: src-tauri/src/services/dictionary/poe2_bases.rs
use std::collections::HashMap;

pub fn get_poe2_canonical_pairs() -> Vec<(String, String)> {
    let mut pairs = Vec::new();
    let static_pairs = [
        // Waystones (尋路石)
        ("尋路石", "Waystone"),
        ("銘刻尋路石", "Inscribed Waystone"),
        ("塔樓尋路石", "Tower Waystone"),
        // Uncut Gems (未切割寶石)
        ("未切割技能寶石", "Uncut Skill Gem"),
        ("未切割精魂寶石", "Uncut Spirit Gem"),
        ("未切割輔助寶石", "Uncut Support Gem"),
        // Runes (符文)
        ("符文", "Rune"),
        ("太陽符文", "Sun Rune"),
        ("鐵符文", "Iron Rune"),
        ("心靈符文", "Mind Rune"),
        ("石之符文", "Stone Rune"),
        ("風暴符文", "Storm Rune"),
        ("冰川符文", "Glacial Rune"),
        ("沙漠符文", "Desert Rune"),
        ("劇毒符文", "Poison Rune"),
        // Currencies & Gold
        ("金幣", "Gold"),
        ("高階重鑄石", "Greater Orb of Scouring"),
        ("次級重鑄石", "Lesser Orb of Scouring"),
        ("高階點金石", "Greater Orb of Alchemy"),
        ("次級點金石", "Lesser Orb of Alchemy"),
        ("重挫石", "Orb of Annulment"),
        ("神聖石", "Divine Orb"),
        ("崇高石", "Exalted Orb"),
        ("混沌石", "Chaos Orb"),
        ("瓦爾寶珠", "Vaal Orb"),
        ("機會石", "Orb of Chance"),
    ];

    for (zh, en) in static_pairs {
        pairs.push((zh.to_string(), en.to_string()));
    }

    for tier in 1..=16 {
        pairs.push((
            format!("尋路石 (階級 {})", tier),
            format!("Waystone (Tier {})", tier),
        ));
    }

    for tier in 1..=20 {
        pairs.push((
            format!("未切割寶石 (階級 {})", tier),
            format!("Uncut Gem (Tier {})", tier),
        ));
    }

    pairs
}

pub fn get_poe2_item_map() -> HashMap<String, String> {
    let mut map = HashMap::new();

    for (zh, en) in get_poe2_canonical_pairs() {
        map.insert(zh.clone(), en.clone());
        map.insert(en.clone(), en);
    }

    // Shorthand aliases for Waystones and Uncut Gems
    for tier in 1..=16 {
        let en_tier = format!("Waystone (Tier {})", tier);
        map.insert(format!("尋路石 T{}", tier), en_tier.clone());
        map.insert(format!("尋路石 (T{})", tier), en_tier);
    }

    for tier in 1..=20 {
        let en_gem = format!("Uncut Gem (Tier {})", tier);
        map.insert(format!("未切割寶石 T{}", tier), en_gem);
    }

    map
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_poe2_item_map_lookups() {
        let map = get_poe2_item_map();
        assert_eq!(map.get("尋路石"), Some(&"Waystone".to_string()));
        assert_eq!(
            map.get("未切割技能寶石"),
            Some(&"Uncut Skill Gem".to_string())
        );
        assert_eq!(
            map.get("尋路石 (階級 15)"),
            Some(&"Waystone (Tier 15)".to_string())
        );
        assert_eq!(
            map.get("未切割寶石 (階級 19)"),
            Some(&"Uncut Gem (Tier 19)".to_string())
        );
        assert_eq!(map.get("太陽符文"), Some(&"Sun Rune".to_string()));
    }
}
